File: Awful.apk/src/main/java/com/ferg/awfulapp/task/LoginRequest.java
package com.ferg.awfulapp.task;

import android.content.ContentValues;
import android.content.Context;
import android.net.Uri;
import android.widget.Toast;

import com.ferg.awfulapp.constants.Constants;
import com.ferg.awfulapp.network.NetworkUtils;
import com.ferg.awfulapp.preferences.AwfulPreferences;
import com.ferg.awfulapp.thread.AwfulMessage;
import com.ferg.awfulapp.thread.AwfulPost;
import com.ferg.awfulapp.util.AwfulError;

import org.jsoup.nodes.Document;

/**
 * Created by matt on 8/8/13.
 */
public class LoginRequest extends AwfulRequest<Boolean> {
    private String username;
    public LoginRequest(Context context, String username, String password) {
        super(context, null);
        this.username = username;
        addPostParam(Constants.PARAM_ACTION, "login");
        addPostParam(Constants.PARAM_USERNAME, username);
        addPostParam(Constants.PARAM_PASSWORD, password);
    }

    @Override
    protected String generateUrl(Uri.Builder urlBuilder) {
        return Constants.FUNCTION_LOGIN_SSL;
    }

    @Override
    protected Boolean handleResponse(Document doc) throws AwfulError {
        Boolean result = NetworkUtils.saveLoginCookies(getContext());
        if(result){
            AwfulPreferences prefs = AwfulPreferences.getInstance(getContext());
            prefs.setStringPreference("username", username);
        }
        return result;
    }

    @Override
    protected boolean handleError(AwfulError error, Document doc) {
        if(error.networkResponse.statusCode == 302){
            Boolean result = NetworkUtils.saveLoginCookies(getContext());
            if(result){
                AwfulPreferences prefs = AwfulPreferences.getInstance(getContext());
                prefs.setStringPreference("username", username);
            }
            return result;
        }else{
            return error.getErrorCode() == AwfulError.ERROR_PROBATION || error.isCritical();//Don't allow probation to pass
        }
    }
}

File: Awful.apk/src/main/java/com/ferg/awfulapp/task/LoginRequest.kt
package com.ferg.awfulapp.task

import android.content.Context
import com.android.volley.NetworkResponse
import com.ferg.awfulapp.constants.Constants.*
import com.ferg.awfulapp.network.NetworkUtils
import com.ferg.awfulapp.preferences.Keys
import com.ferg.awfulapp.util.AwfulError
import org.jsoup.nodes.Document

/**
 * AwfulRequest that sends a login request to the site, and stores login cookies and sets
 * the current username if successful.
 */
class LoginRequest(context: Context, private val username: String, password: String)
    : AwfulRequest<Boolean>(context, FUNCTION_LOGIN_SSL, isPostRequest = true) {

    init {
        with(parameters) {
            add(PARAM_ACTION, "login")
            add(PARAM_USERNAME, username)
            add(PARAM_PASSWORD, password)
        }
    }

    @Throws(AwfulError::class)
    override fun handleResponse(doc: Document): Boolean = validateLoginState()

    override fun handleError(error: AwfulError, doc: Document): Boolean =
            error.networkResponse?.isRedirect == true || !error.isCritical


    private val NetworkResponse.isRedirect get() = this.statusCode == 302

    /**
     * Check if we've received login cookies, and if so store the username we used to log in
     */
    private fun validateLoginState(): Boolean {
        return NetworkUtils.saveLoginCookies(context).also { success ->
            if (success) preferences.setPreference(Keys.USERNAME, username)
        }
    }

}
